src/osiris_competing_read_coordinator.erl

```diff
@@ -0,0 +1,298 @@
+                            {gb_sets:to_list(InF0) ++ InF, Rs0}
+                    end, {[], Readers0}, Tags),
+    Pending = Pending0 ++ lists:sort(InFlight),
+    SQ = queue:from_list(lists:foldl(fun(Tag, Acc) ->
+                                             lists:delete({Pid, Tag}, Acc)
+                                     end, queue:to_list(SQ0), Tags)),
+                        
+    {noreply, notify_readers(State#state{readers = Readers,
+                                         service_queue = SQ,
+                                         reader_groups = Groups,
+                                         pending = Pending})};
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% This function is called by a gen_server when it is about to
+%% terminate. It should be the opposite of Module:init/1 and do any
+%% necessary cleaning up. When it returns, the gen_server terminates
+%% with Reason. The return value is ignored.
+%% @end
+%%--------------------------------------------------------------------
+-spec terminate(Reason :: normal | shutdown | {shutdown, term()} | term(),
+                State :: term()) -> any().
+terminate(_Reason, _State) ->
+    ok.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Convert process state when code is changed
+%% @end
+%%--------------------------------------------------------------------
+-spec code_change(OldVsn :: term() | {down, term()},
+                  State :: term(),
+                  Extra :: term()) -> {ok, NewState :: term()} |
+          {error, Reason :: term()}.
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% This function is called for changing the form and appearance
+%% of gen_server status when it is returned from sys:get_status/1,2
+%% or when it appears in termination error logs.
+%% @end
+%%--------------------------------------------------------------------
+-spec format_status(Opt :: normal | terminate,
+                    Status :: list()) -> Status :: term().
+format_status(_Opt, Status) ->
+    Status.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+notify_readers(#state{log = Seg0,
+                      service_queue = SQ0,
+                      pending = Pending,
+                      batch_size = BatchSize,
+                      readers = Readers0} = State) ->
+    case queue:peek(SQ0) of
+        {value, Reader} ->
+            #{credit := Credit} = maps:get(Reader, Readers0),
+            case Pending of
+                [] ->
+                    case read_chunks_from_segment(min(BatchSize, Credit), Seg0) of
```

Comment:
if I'm reading this right this would assign all outstanding chunks to the next reader - I think we should just assign one chunk, then if the reader still has credit, put it at the back of the service queue and server the next chunk to any other readers that are waiting